apps/cowswap-frontend/src/widgetEventEmitter.ts

```diff
@@ -1,5 +1,210 @@
+  return !!(
+    payload &&
+    typeof payload === 'object' &&
+    'chainId' in payload &&
+    // An order payload must have either:
+    // 1. An 'order' property (standard case)
+    // 2. A 'transactionHash' property AND be related to an order operation (like cancellation)
+    // This ensures we don't treat a regular transaction as an order payload
+    ('order' in payload ||
+      // For cancelled orders, we might only have the transactionHash
+      // The presence of specific event types (like ON_CANCELLED_ORDER) helps distinguish
+      // order-related transactions from regular transactions
+      ('transactionHash' in payload && payload.transactionHash !== undefined))
+  )
+}
+
+// Specialized helper function for string properties to ensure we always return a string
+const safeGetString = <T, K extends keyof T>(obj: T | undefined, key: K, fallback: string = ''): string => {
+  const value = obj && obj[key] !== undefined ? obj[key] : fallback
+  return String(value || '')
+}
+
+// Map CowWidgetEvents to GTM event names - only include the ones we use
+const EVENT_MAPPING: Partial<Record<CowWidgetEvents, string>> = {
+  [CowWidgetEvents.ON_POSTED_ORDER]: 'order_submitted',
+  [CowWidgetEvents.ON_FULFILLED_ORDER]: 'swap_executed',
+  [CowWidgetEvents.ON_CANCELLED_ORDER]: 'swap_cancelled',
+  [CowWidgetEvents.ON_EXPIRED_ORDER]: 'swap_expired',
+}
+
+// Generate default tracking payload
+const createDefaultTrackingPayload = (): OrderTrackingPayload => {
+  return {
+    walletAddress: '',
+    orderId: '',
+    chainId: 0,
+    sellToken: '',
+    buyToken: '',
+    sellAmount: '',
+    buyAmount: '',
+  }
+}
+
+// Helper to extract common order properties
+const getCommonOrderProperties = (payload: unknown, isPostedOrder = false): OrderTrackingPayload => {
+  if (isPostedOrder && isPostedOrderPayload(payload)) {
+    return {
+      walletAddress: payload.owner,
+      orderId: payload.orderUid,
+      chainId: payload.chainId,
+      sellToken: safeGetString(payload.inputToken, 'address'),
+      buyToken: safeGetString(payload.outputToken, 'address'),
+      sellAmount: payload.inputAmount?.toString() || '',
+      buyAmount: payload.outputAmount?.toString() || '',
+    }
+  }
+
+  if (isOrderPayload(payload)) {
+    return {
+      walletAddress: safeGetString(payload.order, 'owner'),
+      orderId: safeGetString(payload.order, 'uid'),
+      chainId: payload.chainId,
+      sellToken: safeGetString(payload.order, 'sellToken'),
+      buyToken: safeGetString(payload.order, 'buyToken'),
+      sellAmount: safeGetString(payload.order, 'sellAmount'),
+      buyAmount: safeGetString(payload.order, 'buyAmount'),
+      ...(payload.transactionHash ? { transactionHash: payload.transactionHash } : {}),
+    }
+  }
+
+  // Fallback for unexpected payload types
+  return createDefaultTrackingPayload()
+}
+
+// Helper to set up an event handler with common logic
+const setupEventHandler = <T extends CowWidgetEvents>(
+  event: T,
```

Comment:
You already have event here which is a known value from `CowWidgetEvents` enum.
You don't need `isPostedOrder` flag and `isPostedOrderPayload`, `isOrderPayload` as well.

Reply:
@fairlighteth let me know if I can help with that